### PR TITLE
Custom auth in failure function

### PR DESCRIPTION
**Pull request by CahidArda**
Run [custom auth](https://upstash.com/docs/workflow/howto/security#custom-authorization-method) in failureFunction by default.

also fixes:
- we checked retries like `(retries)`, so when it was set to 0, it wasn't being set properly.
- when a context.call is made to an endpoint which returns no body

### CODE REVIEW

File: examples/ci/app/ci/upstash/redis.test.ts [hunk 1]
```diff
@@ -80,4 +80,21 @@ describe("redis", () => {
...
+  test("should fail if marked as failed", async () => {
+
+    const route = "fail-route"
+    const randomId = `random-id-${nanoid()}`
+    const result = `random-result-${nanoid()}`
+
+    // increment, save and check
+    await redis.increment(route, randomId)
+    await redis.saveResultsWithoutContext(route, randomId, result)
+    await redis.checkRedisForResults(route, randomId, 1, result)
```
Comment (fahreddinozcan) [inline examples/ci/app/ci/upstash/redis.test.ts:93]: Why do we check this? We already know it'd work

Comment (CahidArda) [reply examples/ci/app/ci/upstash/redis.test.ts:93]: just a checkpoint to make sure that the state is set in redis, and only difference is the `failWithoutContext` method

File: examples/ci/app/ci/upstash/redis.ts [hunk 2]
```diff
@@ -80,6 +80,39 @@ export const saveResult = async (
...
 
+export const failWithoutContext = async (
+  route: string,
+  randomTestId: string
+) => {
+  const key = getRedisKey("fail", route, randomTestId)
+  const pipe = redis.pipeline()
+  pipe.set<boolean>(key, true)
+  pipe.expire(key, EXPIRE_IN_SECS)
+  await pipe.exec()
```
Comment (fahreddinozcan) [inline examples/ci/app/ci/upstash/redis.ts:91]:
```suggestion
  await pipe.set<boolean>(key, true, {ex: EXPIRE_IN_SECS})
```

File: examples/ci/app/ci/upstash/redis.test.ts [hunk 3]
```diff
@@ -80,4 +80,21 @@ describe("redis", () => {
...
+
+    // mark as failed and check
+    await redis.failWithoutContext(route, randomId)
+    expect(redis.checkRedisForResults(route, randomId, 1, result)).rejects.toThrow("Test has failed because it was marked as failed with `fail` method.")
```
Comment (fahreddinozcan) [inline examples/ci/app/ci/upstash/redis.test.ts:97]: We can maybe hardcode this error message, to make it less likely to be broken by string mutation. It'd just be easier to manage.

File: examples/ci/app/test-routes/auth/custom/workflow/route.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    method,
+    headers: {
+      [ CI_RANDOM_ID_HEADER ]: randomId,
+      [ CI_ROUTE_HEADER ]: route,
+      "Upstash-Workflow-Is-Failure": "true"
+    }
+  })
+
+  expect(status, expectedStatus)
+  console.log("BODE", JSON.stringify(body));
```
Comment (fahreddinozcan) [inline examples/ci/app/test-routes/auth/custom/workflow/route.ts:53]: We may remove this?

File: examples/ci/app/test-routes/auth/custom/target/route.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { WorkflowContext } from "@upstash/workflow";
+import { serve } from "@upstash/workflow/dist/nextjs";
```
Comment (fahreddinozcan) [inline examples/ci/app/test-routes/auth/custom/target/route.ts:2]: why `/dist/nextjs`

Comment (CahidArda) [reply examples/ci/app/test-routes/auth/custom/target/route.ts]: wrong import, fixed.

**Review by fahreddinozcan — APPROVED** [review]

**Pull request merged**